### PR TITLE
cody: cleanup the design of the VSCode History view

### PR DESCRIPTION
**Pull request by toolmantim**
As a warmup to starting work on the VSCode Extension, I did some cleanups of the design of the Cody VSCode chat history. This pushes the most important content in the history panel (the history items themselves) much higher, and moves the clear history to just a button next to the header (instead of a whole section).

| Before | After |
| - | - |
|  Screenshot 2023-04-28 at 5 17 40 pm | Screenshot 2023-04-28 at 5 18 27 pm |

Notes:

* In the process of battling the default button styles, I switched it to use standard VSCodeButton elements, which give us a bunch of nice defaults for free (hover etc).
* There's probably more work that can be done in follow up PRs to bring it up to speed with #51072, and any new work in #51165, but this at least improves what we have now.
* There was a keyboard handler and tabindex in there, but I couldn't seem to get it to trigger. Would love help for adding it back in if it was working.

## Future Fixes / Discussions

1. When starting your first chat, and then clicking history, I was expecting it to appear in Chat History. But it's more like "Previous Chat Sessions" — you have to reload, or leave and come back, to have new things appear. I feel like the current live session could still be there at the top of the History (if there's been a message sent), and clicking it would just go back to it.
2. When clicking on a recipe, I thought it would start a new chat, and the old chat would end up in history. But it just adds to the current chat 🤔
3. If you use fixup, the chat messages (and therefore chat history buttons) are a bit empty and awkward. Screenshot 2023-04-28 at 5 49 23 pm Screenshot 2023-04-28 at 5 51 04 pm<br> Perhaps it could say "I've updated your code." instead of "Here is my attempt at rewriting the selected code:"
4. Sometimes the history buttons contain escaped HTML, which would normally be rendered as HTML in the normal chat. It might be nice to fix that.

## Test plan

Would appreciate another person’s manual test of this, given it's my first contribution to the extension 🙏🏼

* With no history, open Chat History view, verify it works okay, and that Clear History button is disabled
* Perform a chat, reload the window, open Chat History view, verify buttons and labels
* Perform another chat, reload the window, open Chat History view, verify buttons and labels
* Click Clear History

### CODE REVIEW

**Review by eseliger — APPROVED** [review]
nice 🚀

**Review by SuperAuguste — APPROVED** [review]
I love it! Awesome design Tim!!

File: client/cody/webviews/UserHistory.module.css [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+.clear-button-container {
+    margin-left: auto;
+}
+
+.items-container {
+    margin: 0 1rem;
+    display: flex;
+    gap: 1rem;
+    flex-direction: column;
```
Comment (abeatrix) [inline client/cody/webviews/UserHistory.module.css:20]: We might need to add overflow and wrap to the item container:
image

File: client/cody/webviews/UserHistory.tsx [hunk 2]
```diff
@@ -40,57 +40,53 @@ export const UserHistory: React.FunctionComponent<React.PropsWithChildren<Histor
...
                             )
-                        })}
+                            .map(chat => {
+                                const lastMessage = chat[1].interactions[chat[1].interactions.length - 1].assistantMessage
+                                if (!lastMessage?.displayText) {
+                                    return null
+                                }
+
+                                return (
+                                    <VSCodeButton
```
[the remark this answers is not in the document]
Comment (abeatrix) [reply client/cody/webviews/UserHistory.tsx:72]: Here is a screenshot that shows the weird line in each buttons as reported previously in https://github.com/sourcegraph/sourcegraph/pull/50960 :
image

**Review by abeatrix — CHANGES_REQUESTED** [review]
Apart from the overflow issues as mentioned, I really like the new design, def much better than what we have now!

File: client/cody/webviews/UserHistory.module.css (hunk 1 above)
Comment (toolmantim) [reply client/cody/webviews/UserHistory.module.css:20]: Ooh good catch! Thanks @abeatrix. I've just updated it back to using our own `<button>` (styling the shadow root element is impossible), and fixed recipe buttons while I was there.

I couldn't replicate the wrapping problem though. Could you check this PR again with your local data? Thank you!

**Comment by abeatrix** [comment]
I'm still seeing the overflow issue, maybe it's just me? 🤔 
image

**Comment by abeatrix** [comment]
Looks like it's fixed, we will need to update the dependencies for this: https://github.com/microsoft/vscode-webview-ui-toolkit/issues/384

File: client/cody/webviews/UserHistory.module.css [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    Sadly we have to use our own button instead of VSCodeButton because it has a
+    #shadow-root style with max-width: 300px AND we get a weird border when the
+    window is wide. So we use a <button> and set our own styles (mostly found by
+    inspecting the standard button)
+**/
+.item-button {
+    all: unset;
+    box-sizing: border-box;
+    display: inline-flex;
+    justify-content: center;
```
Comment (abeatrix) [inline client/cody/webviews/UserHistory.module.css:33]:
```suggestion
    justify-content: flex-start;
    width: 100%;
```
To fix centered text issue:

image

**Review by abeatrix — APPROVED** [review]
I've updated the package that has the fix for the max-width issue and looks like it works with your changes too.
The only issue is the `justify-content: center;` I mentioned above but other than that i think it's good to go :D 

Here is what it looks like on my side after updating from `center` to `flex-start` with width set to 100%:
image

image

**Comment by toolmantim** [comment]
Amazing, thank you @abeatrix! I'll get this cleaned up and ready to go.

**Comment by toolmantim** [comment]
This is ready to go

**Pull request merged**